keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1347 @@
+
+Second, In order to mitigate this risk we are going to document and define all
+conditions as constants in the k8s.io/apis/core/v1 package. As a result, every
+addition of a new reason will require an API review. The constants will allow
+users of the package to reduce the risk of typing mistakes.
+
+#### Garbage collected pods
+
+The Pod status (which includes the `conditions` field and the container exit
+codes) could be lost if the failed pod is garbage collected.
+
+Losing Pod's status before it is interpreted by Job Controller can be prevented
+by using the feature of [job tracking with finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/).
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### New PodConditions
+
+The already existing `status.conditions` field in Pod will be used by kubernetes
+components (such as: kube-scheduler, kubelet pod-gc) to append a dedicated
+condition when they send the delete operation.
+
+The API call to append the condition will be issued as a pod status update call
+before the Pod delete request. This way the Job controller will already see the
+condition when handling a failed pod.
+
+During the implementation process we are going to review the places where the
+pod delete requests are issued to modify the code to also append a meaningful
+condition with dedicated `Type`, `Reason` and `Message` fields based on the
+invocation context.
+
+### JobSpec API
+
+We extend the Job API in order to allow to apply different actions depending
+on the conditions associated with the pod failure.
+
+```golang
+// BackoffPolicyAction specifies how a Pod failure is handled.
+type BackoffPolicyAction string
+
+const (
+
+	// This is an action which might be taken on pod failure - mark the
+	// pod's job as Failed and terminate all running pods.
+	BackoffActionTerminate BackoffPolicyAction = "Terminate"
+
+	// This is an action which might be taken on pod failure - the pod will be
+	// restarted and the counter for .backoffLimit will not be incremented.
+	BackoffActionIgnore BackoffPolicyAction = "Ignore"
+)
+
+type BackoffPolicyOnExitCodeInterval struct {
+	MinInclusive int
+	MaxInclusive int
+}
+
+type BackoffPolicyOnExitCodeRequirement struct {
+	// Restricts the check for exit codes to only apply to the container with
+	// the specified name. When empty the rule applies to all containers
+	// +optional
+	ContainerName *string
+
+	// Specifies the set of intervals. The requirement is satisfied when the
+	// actual exit code is in any of the intervals
+	// +optional
+	InRange []BackoffPolicyOnExitCodeInterval
```

Comment:
I prefer we have an operator field and a single `values []int` to be closer to node affinity.

Reply:
I'm happy with going this path if there is such consensus.

I was a little bit concern that by making it closer to (but not exactly like) the node affinity we can make it confusing. The syntax would not make use of the "key" field as there. Also, a set of non-standard operators like InClosedRange.